### ISSUE TITLE
Allowing `infer` after `extends T` in conditional types would simplify complicated type definitions massively

### ISSUE THREAD

**Issue by nth-commit** [microsoft/TypeScript#47330]
# Suggestion

## 🔍 Search Terms

- distributed conditional types
- infer keyword

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Doing this would allow the complete removal of an entire branch, and it's much more intuitive (the hanging `infer` stumped me a whole lot when I was learning).

Before:

```typescipt
InferringType<T> = T extends infer U ? U extends SomeConcreteType ? U : never : never
```

After:

```typescipt
InferringType<T> = T extends SomeConcreteType infer U ? U : never
```

### Alternatives

1. Something consistent with the [`as` keyword ](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-1.html#key-remapping-in-mapped-types) in template string literal types e.g.

```typescipt
InferringType<T> = T extends SomeConcreteType as U ? U : never
```

2. No extra syntax, just smart control flow analysis on the type variable.

```typescipt
InferringType<T> = T extends SomeConcreteType
  ? T // T extends SomeConcreteType on this branch
  : never // T does not extend SomeConcreteType on this branch
```

## 📃 Motivating Example

[This playground link](https://www.typescriptlang.org/play?#code/PQKhCgAIUh1ALAhgFwOQGdIHMD2BLAOy0hwMngFMAnCgfihgcgEFJkBPABwoFp1EAZhUgBbHABMKAGxIDIFAG4UCyTOgCuAI3QBjKnk7I8pADSRNidBXEky8HAHdIAJWt5MABS0BlLZAc4VADW6AB0kACSZMhIyKIS0mbsOOqQOohkGtp6eJrCyDiQGfJKKubsTCIoOvCExMUUAB6IOnHoyPpE4QDyVEkpacWI4jaI-nhS4ulUNgWi1fDzBOyQQYTimDhyisqqoUxMACq1mHginFIUIrsoxmSkUitSODghkAKBReOT06NxMcJlDYtmxKNg8KVIO1OsQABQUUJYcLqKxUACipWQoH8gRCZhR1Axu1AAB1gOgcNcYnU0VIrEwAgRUHEAsEAJT7aDAcA80AQaCQQ5cXgrZAUdroACEDG54EkOikiBoaVI7RKu3QAC5IESVOgeTs9aEsrp9HlYfC2ZAALwAPkgAG8oCqCGqKAAGbW6uLW+SQYDAFjLdUqcAAXzZBsxYRNOXNqAJVGAFgIBGsqDMlpt9qdkBdbs9kAAqqjvaES9QAEIZNPib02v0BnWYyDuIrodB4LAERCaS5sQoA8w16wh5DOnSquIUACM2or6Mx5dRAGEcO11NcV0giNZ677hE20VQqIFtULuJBUAvq6m95jUK3MAQcHFLJ3u73+3MOJfr6v12QTcKG3DIsHvXZUE5fNpwAJm1DwlR0aR90bQNj1PKhz2FK8bxHOsHyfSAXzfDsux7Pt8kHHDUEQqhkKkb0oPDSNwENPZYzNChYQTVFQAzSAsztR0JyneRCwXVDD3Qls23fcivyolgADkAE1IETMdRNdac50gOiGKk-10JPM9BRovC7wI3ZIAAH2LACNy3HdwOslRHzbEj2w-Cjv2ov8DJQh99gjKMNWNLRTVybjUBI5gCCYzMKCtYTcxg8TtTTJQqAbaTmxstsAAMsuoQqzAcWoaiIsZJ3OCZAVMnK8DkZJUg6FYMhscRCgydhqSIAdW3HUK2OjCLsi4ni4oSzF+KSlKc20gtMp2HKD2M-KyiKkqqEKoosEQQgWJ5cA+SYCJzkua4VFuUgZQNRpOECf4cOYbQOhaZBvQAHgvQFGjFAgNihDo6jMRD2GeYZ7V9NLfwobDuGdPBxG1aE6mdThEEhnBhgQ7GofEcMHqeqgXsvSSW19SndmXKt8PreyaZUOmqDXJyQJciDQzYx7nuIxBrnQLHkIcwkWzh173qoT7meQX6cKaQHgfRohwYJ3HxBhlhpc+n7CsTYAABIHT+sMyv0jXoZ5PMmlJ8nhEs2tULe6FZdLTFvudPNUBTWsM29kS82D-xECkS5kDR0GiEDsNnVtG35D5sm2Bwhd2aA5ywO5n0dbd1o5a9kPUEnDnQN3cQA5DtKQ4cMOI6jmFY-j4neft1PL0Cxiqct+igtprvEOQGoc7s3vDKXLvvGQFAKCHke3PHNv+Z7IWReELv60ly9XY+1pN89v6k+VtRo6wdWcehhtd5l1p9axvupGN03hXNi-CYT507f5+Hx+kefKCL2vrrfeSF+4qG+qgLGw9AECQdGGT+ttk4Oz-lIaes8AGj19DfT6B9diQPaBghY6YzDwITnHZeKdf6oTlmPPBPNf5RCEIuXYAAxU8Ig-oAFkFgK24DwmBx8gSnxhNrbhCwhFA1OAQZhkQZHUBoOIcRMDnS0DkcwxRyiqpK2EZtZAABtZkwpUAAF1A5qLRADW+8tvRmBwXfT2TCFHWC0fAMwvVbQJxDtqJxJ4XHCgEdogGujCom0ILI2AEwphKkkEogJCwwwgEKuYvRkjgbhOoHolclgKCBzzBYzE2SrBpMwA6DuCNIChIdJEn4MT-H8ISSbXq5tIBxxDiHApuwim5PacHFa2U8mQH6dQQOwyqDOjGSTH+adXSRTjBQVh6gCCtDuA2WEC1IAKHwETShKDvSYFhpjLQUh3DwHhJiL0mI2Tai2SjZ0nFoqwnSOHCwOggjanObsS5uwNm3PENc4ssyJrRUWcsowd08wPLyHwiggTFg6KkSDURsJA7w0RrChYJhA7PKkK895gl2I+PkSwlQ7DKSuJhXC20vztnOgBUWIFUU8igpWXdOOQA) - has some documentation that should provide some context for the following code, but this is a pretty universal pattern anyway.

Before:

```typescript
type InferEventFromTypeMatch<TypeMatch extends string> = TypeMatch extends infer InferredTypeMatch
  ? InferredTypeMatch extends Event['type']
    ? Extract<Event, AbstractEvent<InferredTypeMatch, any>>
    : InferredTypeMatch extends `${infer WildcardedTypeMatch}*`
    ? Event extends infer EventCase
      ? EventCase extends { type: `${WildcardedTypeMatch}${any}` }
        ? EventCase
        : never
      : never
    : never
  : never
```

After:

```typescript
type InferEventFromTypeMatch0<TypeMatchString extends string> =
  TypeMatchString extends Event['type'] infer ExactTypeMatchString // Can give a much more meaningful label
    ? Extract<Event, AbstractEvent<ExactTypeMatchString, any>>
    : InferredTypeMatch extends `${infer WildcardedTypeMatch}*`
      ? Event extends { type: `${WildcardedTypeMatch}${any}` } infer WildcardTypeMatch // Can give a much more meaningful label here too
        ? WildcardTypeMatch 
        : never
      : never // Only two hanging nevers, rather than four
```

**Comment by webstrand** [comment]
I think what you're looking for already exists, with a slightly different syntax:
```ts
type Is<T extends U, U> = T
```

Couldn't follow the logic of your type, so I may have made some errors. 
```ts
type Is<T extends U, U> = T;

type InferEventFromTypeMatch<TypeMatchString extends string> =
  TypeMatchString extends Is<infer ExactTypeMatchString, Event['type']>  // Can give a much more meaningful label
    ? Extract<Event, AbstractEvent<ExactTypeMatchString, any>>
    : TypeMatchString extends `${infer WildcardedTypeMatch}*`
      ? Event extends Is<infer WildcardTypeMatch, { type: `${WildcardedTypeMatch}${any}` }>  // Can give a much more meaningful label here too
        ? WildcardTypeMatch 
        : never
      : never // Only two hanging nevers, rather than four
  ```
  
  But in your specific example you don't need it at all:
  ```ts
  type InferEventFromTypeMatch<TypeMatchString extends string> =
  TypeMatchString extends Event['type']  // Can give a much more meaningful label
    ? Extract<Event, AbstractEvent<TypeMatchString, any>>
    : TypeMatchString extends `${infer WildcardedTypeMatch}*`
      ? Event extends { type: `${WildcardedTypeMatch}${any}` }  // Can give a much more meaningful label here too
        ? Event 
        : never
      : never // Only two hanging nevers, rather than four
  ```
  There's a cost to inference, so idiomatic types do not bother with it.

**Comment by RyanCavanaugh** [comment]
This is already how it works?
```ts
type Testable = { test(): void };
type GetTest<T extends Testable> = { m: T };
// Nothing needed here to match the constraint
type Foo<T> = T extends Testable ? GetTest<T> : never;
```

A more straightforward motivating example would be helpful here.

**Comment by RyanCavanaugh** [comment]
@yudai-nkt I think this a misunderstanding - almost certainly, we wouldn't change the distributivity behavior under this construct

**Comment by yudai-nkt** [comment]
@RyanCavanaugh I don't want the distributivity behavior to be changed. Let me rephrase my issue and possible solutions.

My original `DottedKey` implementation has 4 nested conditional types and I want to reduce the number to 3 or less. My  workaround in my previous comment would have been reduced it to 3, but it is not working unfortunately.

```typescript
type DottedKey<T, K extends keyof T = keyof T> = K extends string
  ? [T[K], T[K]] extends [Record<string, any>, infer V]
    ? V extends (infer E)[]
      ? K | `${K}.${number | "length"}` | `${K}.${number}.${DottedKey<E>}`
      : K | `${K}.${DottedKey<V>}`
    : K
  : never;
```

The feature requested by the issue author will also reduce the number of conditional types. `[T[K], T[K]] extends [Record<string, any>, infer V]` will be replaced with something like `T[K] extends Record<string, any> infer V`. 

In both cases, the nest depth is reduced to 3 and I still have a type parameter `V`, which can be distributive in a conditional type (unlike `T[K]`).

**Comment by RyanCavanaugh** [comment]
Sure, but nothing is going to make `DottedKey` not fail on at least some types.
```ts
type M = { next: M | { data: number } };
type F = DottedKey<M>;
```

**Comment by yudai-nkt** [comment]
Yes, but `M` doesn't need to be covered by `DottedKey` for my specific use case. So I don't really care at the moment.

**Comment by Andarist** [comment]
q: is there any info anywhere on what `T extends infer R ? R : never` does and how it affects other parts of the inference algorithm? Intuitively I've assumed that it's forcing the compiler to resolve T based on the information it was able to gather. 

I've assumed that in a situation like this it wouldn't have any effect on how typechecking works for the argument that should be typechecked with the outcome of `ComputeStuff`:
```ts
interface TypegenDisabled { "@@xstate/typegen": false; }
interface TypegenEnabled { "@@xstate/typegen": true; }

type TypegenConstraint = TypegenEnabled | TypegenDisabled;

type ComputeStuff<TTypesMeta> = TTypesMeta & { /* do stuff */ };

declare function createMachine<
  TTypesMeta extends TypegenConstraint = TypegenDisabled
>(
  config: { tsTypes?: TTypesMeta },
  implementations: TTypesMeta extends infer EvaluatedTypesMeta
    ? ComputeStuff<EvaluatedTypesMeta>
    : never
): void;
```

Unfortunately, it changes the behavior "down the road" but since `tsTypes` is an optional property I also need to "resolve" it quickly to `TypegenDisabled` case, and while this trick works for that, some other test cases of mine break due to the mentioned change of behavior.

This is just an example to illustrate what I'm doing but I'm really after "fixing" it - I would like to better understand how I could influence the order in which TS infers stuff. Basically, I know where the information I need is in the input arguments but I'm not sure how to best control what should be "deferred" by TS. Note that I'm aware of `NoInfer`, `LowInfer`, and other stuff but this is still just a magic box that I find hard to manage in a complex scenario.

**Comment by yudai-nkt** [comment]
I'm trying 4.7 beta and it looks like #48112 fixes this issue although recursion limit is not relaxed yet.

**Comment by theonlypwner** [comment]
I'm trying to understand the motivation behind this issue and #48112, whose [announcement has an example](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#extends-constraints-on-infer-type-variables) of 3 equivalent declarations, the last one of which is added by the pull request:

```ts
type FirstIfString<T> =
    T extends [infer S, ...unknown[]]
        ? S extends string ? S : never
        : never;
```
```ts
type FirstIfString<T> =
    T extends [string, ...unknown[]]
        // Grab the first type out of `T`
        ? T[0]
        : never;
```
```ts
type FirstIfString<T> =
    T extends [infer S extends string, ...unknown[]]
        ? S
        : never;
```

Isn't that also equivalent to this?
```ts
type FirstIfString<T> =
    T extends [infer S, ...unknown[]]
        ? (S & string)
        : never;
```

---

For the example in the first post of this issue, what's the difference between `U extends SomeConcreteType ? U : never` and `U & SomeConcreteType`?

Before:
```ts
type InferringType<T> = T extends infer U ? U extends SomeConcreteType ? U : never : never
```

After:
```ts
type InferringType<T> = T extends infer U ? (U & SomeConcreteType) : never
```

And in the Motivating Example, isn't it equivalent to this?
```ts
type InferEventFromTypeMatch<TypeMatch extends string> =
  TypeMatch extends Event['type']
    ? Extract<Event, AbstractEvent<TypeMatch, any>>
    : TypeMatch extends `${infer WildcardedTypeMatch}*`
      ? Event & { type: `${WildcardedTypeMatch}${any}` }
      : never
```

**Comment by Andarist** [comment]
@RyanCavanaugh I think this is basically solved by the mentioned https://github.com/microsoft/TypeScript/pull/48112 - the only difference is in the allowed order between `infer` and extra `extends`. The OP's example can be rewritten (using the added capability) as follows:
```ts
type InferEventFromTypeMatch<TypeMatchString extends string> =
  TypeMatchString extends infer ExactTypeMatchString extends Event["type"]
    ? Extract<Event, AbstractEvent<ExactTypeMatchString, any>>
    : TypeMatchString extends `${infer WildcardedTypeMatch}/*`
    ? Extract<Event, AbstractEvent<`${WildcardedTypeMatch}/${any}`, any>>
    : never;
```

[TS playground with the full thing](https://www.typescriptlang.org/play?ssl=97&ssc=13&pln=92&pc=1#code/PQKhCgAIUh1ALAhgFwOQGdIHMD2BLAOy0hwMngFMAnCgfigmkgEFJkBPABwoFp1EAZhUgBbHABMKAGxIDIFAG4UCyTOgCuAI3QBjKnk7I8pADSRNidBXEky8HAHdIAJWt5MABS0BlLZAc4VADW6AB0kACSZMhIyKIS0mbsOOqQOohkGtp6eJrCyDiQGfJKKubsDKIoOvCExMUUAB6IOnHoyPpE4QDyVEkpacWI4jaI-nhS4ulUNgVVyDVVBOyQQYTimDhyisqqoQyVACq1mHginFIUIrsoxmSkUitSODghkAKBReOT06NxMcJlDYtmxKNg8KVIO1OsQABQUUJYcLqKxUACipWQoH8gRCZhR1Axu1AAB1gOgcNcYnU0VIrJUAgRUHEAsEAJT7aDAcDgUCMGCHLi8FbICjtdAAQgY3PAkh0UkQNDSpHaJV26AAXJAiSp0ABuHk7XWhLK6fR5WHwtmQAC8AD5IABvKDKgiqigABi1OriNvkesgwGALGWapU4AAvmyDeAjXtTTkLQAiAlUYAWAgEaxJsxW20O52QV3ur2QACqqJ9oQr1AAQhks+Ifbb-YHg833EV0Og8FgCIhNJc2IUAeYG9Yw8gXToVXEKABGLU19GY6uogDCOHa6mu66QRGszb9FADQe1VCogS1gu4kFQy-rmcPmNQkE7BBwcUsPb7A6Hcw4W97w3LdkB3Cg9wyLBn12VBOWLOcACYtQ8RUdGkI9WzPNELyvSAb2EYC63HJsXzfTAPy-bte37Qd8hHIU71Qqh0KkH04MjaNDUxMIE3NChYRTVFQBzSA83tJ1p1neRS2XTCTzbbVMXIrsf1o-9CmYAA5ABNSBU0nKS3TnRdIGY1j5NPdtcKoa9GKIqhH0bZsAB9yxA7dd33aDSNglTKNUmi-3oth7PMjCX32KMYzjXitDNXIBKTSjmAIH1RPEgsjJLLUsyUKgWwU7DlM7AADPLqFKswHFqRZOzGGdzgmQEbLfORklSDoVgyGxxEKDJ2GpIhhzfKdou49UTXixMkpStLMRE3MKGtCTCwQmTcp2ArjyspTdhU8qttKoosEQQhOJjPlKgic5LmuFRblIaVDUaThAn+RjmG0DoWmQH0AB4CPkRpRQIDYoQ6OozFQ9hnmGB0-TWwCKDs7gDSLPBxC1aE6nRyBOEQWGcGGFDCbh8QDQjGLXve0Lbzk5SbRdNyGd2NdiKfXzwyLFnK1XZdN08iDvJglRqbeqg4n7a50AJ9D3MJZSkc+76qF+1mVEBximlB8GcaIaGyeJ8QEZYVXfoBl0i1K1NgAAEkdAiIyqq2zKN4YXTtGMiyaCWPvp1EnNF30zehdW+d2f7XaTDNGxzV21qLIsHEQKRLmQbHIaIPGiwjT3veBv26eEAXQPAyCDy5kOvrD1oNeQKOk8gJMZyFiufPjpvE6TlO04oDOIZhHPIDzosvcjF6i+Rt2WIi-amZ5meLNXcKpFQhZKCr5ml7nlRQlX7xkBQCh15qYPxdp6WxTl4RV+bZXbxrn7WjvzEtdvHWgTULOsENon4ZbE-NWrRLZN1KgTWeUh7aOyFM7EwrsYbk3zjyH2NNJbFx3mvaom9MJAN+q-SOSYCYb2zGYR0EZx4ul9rTaeB8j6ilPjgxmodn7IAIZrJM7Rj6MNIU6ChlNJ40MYpheukA3LsOQDGaeUQhArl2AAMUvCIAiABZbB78KBqI3ofGEwNdbfxhAjF0qjsE6LqHor+b4CCyO1M0VoJjtE-wsWDTAPoADaSZkZJgALqu1oLY1h-0fRmDwSAt+aI7HIAcTUMxBsijLDtHaV2qNNGmKcZ-FxkBSoO0IDY2AEwpiKkkOIaJ8AIygFKn4gJwCG7BJYTUgG2THT5J+EU6wpTykOwGnA+J7BEnJMgBVKgF90HTzLG6aa-F5HqAIK0O4LZYQrQdAofAFNBGjOETxFsa1OBaCkO4eA8JMTekxGyLUKzMZ4z4olWE6Q04WB0EELURzdgnN2EsyAFzxBnPLBM7IUyZlzNIFcyZiUNFaMWBkvWP87SwldsjFJEL4DwKbncqQDynliTjFqGRisFFKNKeC7BdoPlfJdD88Z1y8jTNmUYYFkYDRAA)

**Comment by theonlypwner** [comment]
> ```ts
> type InferEventFromTypeMatch<TypeMatchString extends string> =
>   TypeMatchString extends infer ExactTypeMatchString extends Event["type"]
>     ? Extract<Event, AbstractEvent<ExactTypeMatchString, any>>
>     : TypeMatchString extends `${infer WildcardedTypeMatch}/*`
>     ? Extract<Event, AbstractEvent<`${WildcardedTypeMatch}/${any}`, any>>
>     : never;
> ```

How is this different from my example?

```ts
type InferEventFromTypeMatch<TypeMatch extends string> =
  TypeMatch extends Event['type']
    ? Extract<Event, AbstractEvent<TypeMatch, any>>
    : TypeMatch extends `${infer WildcardedTypeMatch}*`
      ? Event & { type: `${WildcardedTypeMatch}${any}` }
      : never
```

Also, OP's example didn't have the extra `/` characters in the strings.

**Comment by theonlypwner** [comment]
> Sure, but nothing is going to make `DottedKey` not fail on at least some types.
> 
> ```ts
> type M = { next: M | { data: number } };
> type F = DottedKey<M>;
> ```

It seems to work if we rewrite to an alternate `DottedKeyA`: ([playground](https://www.typescriptlang.org/play?ts=4.5.4#code/C4TwDgpgBAIg9sYEAmBpCIA8AVANFVKCADyQDtkBnKAawzgDMpsoBeW+p7APjYKNIQK1SsABOASzIBzAFBQoAfmYBtVAF0B5KlABKEAMZwxyTKMkz8AQzIhu8hUtUatQnVIYQxUAGoPHTj6uwlAAFB5eUACiAJQq6v4BToQAPlAABgAkAN6oAL4AdDlkAK4AtgBGkWkARAA2QtLAABY1eelQaVm5hcXlVWK92fCIKOhYUdztiQEAXPxdOflFwwhIaBiYPlPpM1DzZBAAbl6J86gOB8deANyysgD0D1A1DHBwNZ0vb3AFFVZiGqyUCQZgABj4I3W40w2R+82y-zE83MUmkeTy9kez1e70+XR+K1KlS87S+uN+DRkLXxGUJfRJgz+AN2IOg2AAjJC1mNNnD3gikSjxGi8vFMfcnt88eTCUjaRTmQAvIFs5gAJm5ow2WH5cEFAOFFnRX0RViVB36pIlwPA0ChvJAAEEcPhCCRtNQ6CBGMw+N7fTw+O7BCFUTIHMpsGpNB63NR9EYTGYRZYoDY7IkozHgjpwmRPN5YvE9spUhklkNiQNyVSmq0yYsekSrUycg6dS7JtMkvsFhXm+2eZ2cDGdmcCJcoIcTmI7tjpR9Ze9mYDbaDsGCnVroRgXXqDcioOH0TapRTael6dlq6SOrVCXWaV8ryuGQNCkjWXbmBzt+wO3Gfd4SgM0jxPMV1DPHEfgVOUATgld-hVdd2XVf9YGHIDYRAsCjVFU1kMtRkMSxNUAFk+GyacPXmSi0mo5ArGAKxiJrPIoDyO41QAMR3R1MHI7gbigKVsB-KRRBsYAJGYiQ4DIKAJGoEgDAgShKAkE46hAKBkAgCAwHTCgoDAOANIkCodKUgspAkJAClCdUAFYAA4AE4YlQqAeIwwC90E4TRJxQ5SCBIA))

```ts
type DottedKey<T, K extends keyof T = keyof T> = K extends string
  ? T[K] extends Record<string, any>
    ? T[K] extends infer V
      ? V extends (infer E)[]
        ? K | `${K}.${number | "length"}` | `${K}.${number}.${DottedKey<E>}`
        : K | `${K}.${DottedKey<V>}`
      : never
    : K
  : never;

type DottedKeyA<T, K extends keyof T = keyof T> = K extends string
  ? T[K] extends Record<string, any>
    ? T[K] extends (infer E)[]
      ? K | `${K}.${number | "length"}` | `${K}.${number}.${DottedKeyA<E>}`
      : K | `${K}.${DottedKeyA<T[K]>}`
    : K
  : never;


type M = { next: M | { data: number } };
type F = DottedKey<M>; // Type instantiation is excessively deep and possibly infinite.(2589)
type FA = DottedKeyA<M>; // "next"
```

**Comment by Andarist** [comment]
> How is this different from my example?

Sorry, I didn't notice your post. Either way, there are different ways to write this type - the point is (also mentioned by you) that `infer`+`extends` combination (almost like proposed by the OP) is already available.

**Comment by Andarist** [comment]
I think that the recursion limit issue is mostly mitigated by the ability to write tail recursive conditional types:
```ts
type JoinPath<A extends string, B extends string> = [A] extends [never]
  ? B
  : [B] extends [never]
  ? A
  : `${A}.${B}`;

type _DottedKey<
  T,
  Path extends string = never,
  R extends string = never
> = T extends readonly (infer E)[]
  ? _DottedKey<
      E,
      JoinPath<Path, `${number}`>,
      R | JoinPath<Path, `${number | "length"}`>
    >
  : T extends Record<string, unknown>
  ? keyof T extends infer K extends string
    ? _DottedKey<T[K], JoinPath<Path, K>, R | JoinPath<Path, K>>
    : never
  : R;

type DottedKey<T> = _DottedKey<T>;
```
[TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAUg9gSwHYAUCGwAWAeAglCAD2AiQBMBnKC4AJ2QHMAaKAIQONMursYD4oAXigBtXAF0OJclRFIIANwi1xAKChQA-G3VQAXKNaSi07nMXK1G7bl0GABgBIA3rgC+AOhes39gNyqqqCQUAD6ACJwwCRkANIQINi6ACpMuuhYUlxUNPRIDEJQ8kq0aRoASlkyPHkFwsXKqgLCyVXctBBoZHBIADYgUAAUyABmylAAogCUIlZaYZHREHEJSRrrk2UbsIioGDgZmCxOzkgArgC2AEbKvnxbG5UAPjvIh9iHxy7n1+MvAES9UgMLD-O66DR8OxQVombJQcoQADGcFoZGwuUYLDOSAA1kg4AB3JBQ6xQXEJOAjGFtKijcaxWk1RgQ+YRKIxeKJZIiWLiFjwN77D77Fixe4IqAvQV7LAirBivik9YGBq0aHlAJBcDQRac1bJZoLDnLLnYQ1a4LQZIABkKetNq2cUBGcDgBmdVzQtAMmPyUDcAb4AQA9CH1gA9TSBK0wgCM9pNK0Sztd7qgnu9vt4-rcsyDofDGijMZ1MIATImlsnsKm3R6oF6fczc1KM420AAvVWXG60AMF1RhyPR7UhZIAZir+sSVqpUEwZwYEAA8lcAFbBodFqAl1QopA0BdL1cbwrOXRoG0e1loOM37ZQNDlh+Pp8T19vp8AFk-X6fACsf7-k+ABswEgU+ADsEGQU+AAcsFwU+ACcSHIU+cbXhmrIYY+d73jheGQXeL5EcRJFxh+5EUZRv40bRlFAQxjGUeBLGsZRMEcZxlGITxvGUWhAmCSR5bYReomsc+hGSVJ0nlmRcnyQp1HKSpCn0epGkKcx2k6Qp7H6QZCnccZJkKfx5kWQpwnWTZ0kThJuEOVJaATrJLmuW5E5KV53k+Wp-kBT5WnBSFPl6eFEU+UZ0UxT5ZnxQlPlWclKU+XZ6UZW537OTlMVoN+nkFQlRV+aVKVFUFlUZUVYW1TlRVRY1BVFXFrWlUVSWdZVRVpb1tVFVlg2NWgAH5aNY0ASVU2teNFVzZ1401UtvXjQ1a2DeNLVbaN40dXtU3jT1R1zeNA1nUt40jVda1oKBk13WtFIgAY-xoP8DzPaNbjfT9rV-dlAPeUDIO1WD4M5ZDUMxTDsOg-9CMmfDyM6ajaNSRjmOcdjOMUXj+PIYTRP-iTpPbOTFMBkj1PrFTOMM8jTNQyzANs3dHNHVza081NfO9QLjVC6VIsZWLcO0-zUuCzLwty6LCvi0rkvAw5EuCRrrFawTKvo3r8k6yBRtvibGxmzTavEwb1tW1+YtC3zXNs0zVMk3jGPw5DQNuAEQA)
